File: c/leetcode/简单题/35 搜索插入位置/another.c
int searchInsert(int* nums, int numsSize, int target){
    int i = 0, j = numsSize-1;
    int mid;
    while(i <= j){
        mid = (i + j) / 2;
        if(nums[mid] >= target)
            j = mid - 1;
        else
            i = mid + 1;
    }
    return i;
}

//2021.05.18 一大早的，被二分查找给弄自闭了……
//二分查找的变形题？
//以前下面那个必定是返回mid的，因为你是真的在查找一个确切的值，找到的就刚好是mid
//但是这题略微有点不太一样，就是右边的条件是<=，就是你找一个插入位置。这样的情况下必然是能找到一个准确位置的
//而此时，这个准确位置找到时，i和j两者就很重要了。
//算了算了，经典胡言乱语，反正就是比普通的二分要略有变型，多思考一下还是能明白几个条件为什么这么设置的，就是不如原本的清晰罢了。

File: leetcode/二分查找/33 搜索旋转排序数组/another.go
func search(nums []int, target int) int {
	length := len(nums)
	i, j := 0, length-1
	mid := 0
	for i <= j {
		mid = (i + j) / 2
		if nums[mid] == target {
			return mid
		}
		if nums[mid] >= nums[i] {
			//判断i-mid这段数据是否是升序的；之所以加等号，是有可能i=mid，也就是旋转点为nums[1]的情况下，此时左边只有一个元素，右边都是升序；
			if target <= nums[mid] && target >= nums[i] {
				j = mid - 1
			} else {
				i = mid + 1
			}
		} else {
			if target >= nums[mid] && target <= nums[j] {
				i = mid + 1
			} else {
				j = mid - 1
			}
		}
	}
	return -1
}

//参照过题解的思路，但是为什么还没顺序查找速度快……就离谱
//怎么说呢，除了只有两个元素的样例旋转一下成为降序，把我震撼了以外，其它的都是按着预想的思路来的；
//因为有旋转点的存在，划mid一刀，旋转点要么在左边要么在右边（中间也没关系）
//然后肯定一边有序一边无序，那就四个if都判断一下，如果左边有序且target在内，否则遍历右边；如果左边无序且target在内，否则右边；
//就四种情况i、j的变化不一样，难不算难吧，思路不容易想到。